qutebrowser/browser/tab.py

```diff
@@ -0,0 +1,525 @@
+
+    def current_idx(self):
+        raise NotImplementedError
+
+    def back(self):
+        raise NotImplementedError
+
+    def forward(self):
+        raise NotImplementedError
+
+    def can_go_back(self):
+        raise NotImplementedError
+
+    def can_go_forward(self):
+        raise NotImplementedError
+
+    def serialize(self):
+        """Serialize into an opaque format understood by self.deserialize."""
+        raise NotImplementedError
+
+    def deserialize(self, data):
+        """Serialize from a format produced by self.serialize."""
+        raise NotImplementedError
+
+    def load_items(self, items):
+        """Deserialize from a list of WebHistoryItems."""
+        raise NotImplementedError
+
+
+class AbstractTab(QWidget):
+
+    """A wrapper over the given widget to hide its API and expose another one.
+
+    We use this to unify QWebView and QWebEngineView.
+
+    Attributes:
+        history: The AbstractHistory for the current tab.
+
+        for properties, see WebView/WebEngineView docs.
+
+    Signals:
+        See related Qt signals.
+
+        new_tab_requested: Emitted when a new tab should be opened with the
+                           given URL.
+    """
+
+    window_close_requested = pyqtSignal()
+    link_hovered = pyqtSignal(str)
+    load_started = pyqtSignal()
+    load_progress = pyqtSignal(int)
+    load_finished = pyqtSignal(bool)
+    icon_changed = pyqtSignal(QIcon)
+    # FIXME:qtwebengine get rid of this altogether?
+    url_text_changed = pyqtSignal(str)
+    title_changed = pyqtSignal(str)
+    load_status_changed = pyqtSignal(str)
+    new_tab_requested = pyqtSignal(QUrl)
+    shutting_down = pyqtSignal()
+
+    def __init__(self, win_id, parent=None):
+        self.win_id = win_id
+        self.tab_id = next(tab_id_gen)
+        super().__init__(parent)
+        # self.history = AbstractHistory(self)
+        # self.scroll = AbstractScroller(parent=self)
+        # self.caret = AbstractCaret(win_id=win_id, tab=self, modeman=...,
+        #                            parent=self)
+        # self.zoom = AbstractZoom(win_id=win_id)
+        # self.search = AbstractSearch(parent=self)
+        self.data = TabData()
+        self._layout = None
+        self._widget = None
+        self.backend = None
+
+    def _set_widget(self, widget):
+        # pylint: disable=protected-access
+        self._layout = WrapperLayout(widget, self)
+        self._widget = widget
+        self.history._history = widget.history()
+        self.scroll._widget = widget
+        self.caret._widget = widget
+        self.zoom._widget = widget
+        self.search._widget = widget
+        widget.mouse_wheel_zoom.connect(self.zoom._on_mouse_wheel_zoom)
+        widget.setParent(self)
+        self.setFocusProxy(widget)
+
+    @pyqtSlot()
+    def _on_load_started(self):
+        self.data.viewing_source = False
+        self.load_started.emit()
+
+    @property
+    def cur_url(self):
```

Comment:
Hmm, right - I went for properties as those names already existed in the old API, and if they were functions but older code still does `tab.cur_url` it would have silently worked and might get hard to debug. This was quite at the beginning of the refactoring.

I then forgot about this, but I think it indeed makes more sense to make them methods - for consistency too.